### ISSUE TITLE
Root Certificate install fails with code 26352

### ISSUE THREAD

**Issue by wixbot** [wixtoolset/issues#4875]
I have created a WIX installation MSI that installs certificates to machine store. It is supposed to install a root certificate, but the installation fails if the cert already exists in the store.

I'm using Wix 3.9 R2.

This seems to have been a long running issue.  

It was raised back in 2008:
http://windows-installer-xml-wix-toolset.687559.n2.nabble.com/Certificate-install-to-local-machine-fails-with-code-26352-td1121050.html

and again in 2010:
http://windows-installer-xml-wix-toolset.687559.n2.nabble.com/WiX-cannot-install-certificate-Error-26352-installing-certificate-td4864389.html

and other people have been having the problem more recently:
http://stackoverflow.com/questions/26932405/wix-installation-fails-to-install-certificate-to-root-certification-authorities

What (if anything) is the fix for this.  In all those enormous threads, there doesn't seem to be an actual fix at all - it certainly isn't working in Wix 3.9?  Some people seem to suggest that it was working with Wix 3.5?  Do I need to revert back to that version to get this to work?  Or was it never actually fixed at all?  

Can anyone suggest a specific version of Wix which will not cause this issue, or any other workaround.  This is causing a significant headache for my organisation at present - users are having to manually go in and delete the root cert before installing our product - not good...

Thanks in advance for any help...

> Originally opened by `jon34`

**Comment by wahajali** [comment]
@wixbot any update on this?

**Comment by robmen** [comment]
The bug is open and waiting for someone to start working on a solution. Perhaps you would like to start doing so?

**Comment by briandrennan** [comment]
@robmen, who can I, or where should I, ask about the relevant area of WiX? I wouldn't mind giving this a go, if it's in C#. I unfortunately am not very comfortable in C++, but if it's in C#, I think this should be as simple as just checking to see if the certificate is in the store, and if so, moving on?

I'm happy to work on it if someone points me to the appropriate area in the source folder.

Thanks!

**Comment by robmen** [comment]
It'll need to be in C++. Too many problems using managed custom actions in platform custom actions. Take a look at: `src\ext\ca\serverca\scasched\scacert.cpp` and `src\ext\ca\serverca\scaexec\scacertexec.cpp`

**Comment by kirillkovalenko** [comment]
Is this bug expected to be fixed anytime soon?

**Comment by barnson** [comment]
The bug is open and waiting for someone to start working on a solution. Perhaps you would like to start doing so?

**Comment by HimanshuNandu** [comment]
is this resolved?

**Comment by robmen** [comment]
This issue is open and unassigned. That means it is waiting for someone to investigate the root problem, discuss possible solutions to that problem then implement the decided solution.

If you are interested in doing so yourself, our [developer documentation][devdoc] provides a great checklist for getting started.

If you are not interested then you are waiting for someone else to become interested. If this issue has been open for a long time then there probably isn't much interest in this particular issue. In that case, you'll want to consider how to motivate others to fix it for you. This is a pretty [good list of support options][support].

[devdoc]: http://wixtoolset.org/development/
[support]: https://www.firegiant.com/support-options/

**Comment by JasonStephenson** [comment]
I have bumped into this issue and have extracted the failing code path out of dutil.lib. I can consistently reproduce this using a simple.exe (https://pastebin.com/b0RJZANN)

Steps to reproduce:

1. Manually import a CA certificate into the Group Policy or Enterprise **physical** stores within Trusted Root Certification Authority virtual store.
2. Run the aforementioned simple.exe and watch it fail.

The crux of this problem is that if a certificate already exists in these locations then the `::CertAddCertificateContextToStore(CERT_STORE_ADD_REPLACE_EXISTING)` call fails with `ACCESS_DENIED`, even when ran as Administrator. But the call will not fail if executed with `CERT_STORE_ADD_USE_EXISTING`.

Should this be the case? I don't think so. Would it ever be fixed if raised as a MS issue? Unlikely.

Given the following facts:

1. dutil.lib (and therefore simple.exe) imports to the Registry **physical** store of TRCA
2. mmc.exe -> Certificates snapin will import into the Registry **physical** store by default
3. mmc.exe -> Certificates snapin *always* uses CERT_STORE_ADD_USE_EXISTING
* I have confirmed this by analysing the winapi calls that occur during certificate import for:
** A CA that was never present in the stores
** A CA that was present in the Registry (default) store
** A CA that was present in any other physical store
	
I believe it is safe to use `CERT_STORE_ADD_USE_EXISTING` instead of `CERT_STORE_ADD_REPLACE_EXISTING` in **all cases**. However, if one wanted to reduce risk the following (pseudo) could work:
```
If(::CertAddCertificateContextToStore(CERT_STORE_ADD_REPLACE_EXISTING) == ACCESS_DENIED)
   ::CertAddCertificateContextToStore(CERT_STORE_ADD_USE_EXISTING)
```

To this end I have spent most of the weekend and all of today trying to get WiX running with my own dutil.lib code. Sadly, to no avail.I have got as far as compiling `(dutil.lib + wcautil.lib) --> iisca.dll --> iis.wixlib --> WixToolset.Iis.wixext.dll.`

However, I'm at a loss on how to proceed further as if I attempt to reference WixToolset.Iis.wixext.dll from within my .wixproj I get: 

> candle.exe(0,0): error CNDL0307: Either 'Microsoft.Tools.WindowsInstallerXml.AssemblyDefaultWixExtensionAttribute' was not defined in the assembly or the type defined in extension.

Ideally if someone is able to provide guidance on how to proceed I may be able to put this issue to bed once and for all.

Failing that, if someone with a working build system could compile me a test WixIIsExtension.dll with my aforementioned change I could confirm whether my fix works.

Many thanks,
Jason Stephenson

**Comment by JasonStephenson** [comment]
So, an update. I've managed to figure out how to build the correct WixIIsExtension.dll using MSBuild and [this repo.](https://github.com/wixtoolset/wix3). I made my change, then referenced this dll in my project instead of the production dll.

To my surprise the issue was still present. What is more surprising is that even if I change the error log lines they appear unaltered in the MSI log. 

My changes:

```
..\wix3\src\ext\ca\serverca\scaexec\scacertexec.cpp:278
MessageExitOnFailure(hr, msierrCERTFailedAdd, "JST0: Failed to add certificate to the store.");
```

```
..\wix3\src\ext\ca\serverca\scaexec\scacertexec.cpp:294
MessageExitOnFailure(hr, msierrCERTFailedAdd, "JST1: Failed to add certificate to the store.");
```

```
..\wix3\src\libs\dutil\certutil.cpp:286
ExitWithLastError(hr, "JST JST JST JST JST. Failed to add certificate to the store.");
```

And the log output still shows:
```
MSI (s) (08:7C) [20:36:56:099]: Hello, I'm your 32bit Elevated Non-remapped custom action server.
AddMachineCertificate:  Entering AddMachineCertificate in C:\Windows\Installer\MSI1B92.tmp, version 3.11.4516.0
AddMachineCertificate:  Deleting certificate that begin with friendly name: Mimecast Root CA_wixCert_
AddMachineCertificate:  Adding certificate: Mimecast Root CA_wixCert_1
AddMachineCertificate:  Error 0x80070005: Failed to add certificate to the store.
```

Anyone any idea's why this may be the case? @robmen ?

**Comment by rseanhall** [comment]
Please move these questions to the [wix-devs](https://wixtoolset.org/documentation/) mailing list. 

How is `CertAddCertificateContextToStore` telling whether the given certificate matches the existing one? In other words, are you sure that `CERT_STORE_ADD_USE_EXISTING` will always either end up with your certificate in the store or give an error?

The v3 build system sometimes has problems with incremental builds. You should clean and then build again.

Once again, please move this discussion to wix-devs.

**Comment by JasonStephenson** [comment]
1. I have emailed the list from my work address but it is "held until it can be approved". Sadly I'm operating under some  time constraints and given that my email was pertinent to this bug, I thought I would continue the discussion here. Approve my email and I'll happily use the list.
2. I don't know how `CertAddCertificateContextToStore` as I'm not a windows dev and have limited RE capability. I would assume by the certificates thumbprint (hash).
3. Yes I am sure, my simple.exe example demonstrates this if you tweak it to use `CERT_STORE_ADD_USE_EXISTING` instead. A one line change.
4. I don't think it is an incremental build problem as I have nuked my workspace and rebuilt it multiple times. 

Given that:
* I'm using Production Wix dlls + my own altered WixIIsExtension.dll
* I'm seeing no changed to the log output

I can only conclude that it's not as simple as changing dutil.lib and compiling WixIIsExtension. The production dutil.lib must reside in some of the production dlls and it is that that is being called at  runtime during install. I'll do some digging to confirm this.
